File: glean/rts/stats.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

#ifdef OSS
#include <cpp/memory.h> // @manual
#else
#include <common/hs/util/cpp/memory.h>
#endif

#include "glean/rts/stats.h"

using namespace facebook::hs;

namespace facebook {
namespace glean {
namespace rts {

void marshal(
    const PredicateStats& stats,
    size_t *count,
    int64_t **ids,
    uint64_t **counts,
    uint64_t **sizes) {
  const auto n = stats.size();
  auto ids_arr = ffi::malloc_array<int64_t>(n);
  auto counts_arr = ffi::malloc_array<uint64_t>(n);
  auto sizes_arr = ffi::malloc_array<uint64_t>(n);
  size_t i = 0;
  for (const auto x : stats) {
    if (x.second.count) {
      ids_arr[i] = x.first.toThrift();
      counts_arr[i] = x.second.count;
      sizes_arr[i] = x.second.memory;
      ++i;
    }
  }
  *count = i;
  *ids = ids_arr.release();
  *counts = counts_arr.release();
  *sizes = sizes_arr.release();
}

}
}
}

File: glean/hs/Glean/RTS/Foreign/Stats.hs
{-
  Copyright (c) Meta Platforms, Inc. and affiliates.
  All rights reserved.

  This source code is licensed under the BSD-style license found in the
  LICENSE file in the root directory of this source tree.
-}

module Glean.RTS.Foreign.Stats
  ( marshalPredicateStats
  ) where

import Control.Monad (forM)
import Data.Int
import Data.Word
import Foreign.C.String
import Foreign.C.Types
import Foreign.Ptr
import Foreign.Storable

import Util.FFI

import Glean.FFI
import Glean.RTS.Types (Pid(..))
import Glean.Types

marshalPredicateStats
  :: (Ptr CSize
      -> Ptr (Ptr Int64)
      -> Ptr (Ptr Word64)
      -> Ptr (Ptr Word64)
      -> IO CString)
  -> IO [(Pid, PredicateStats)]
marshalPredicateStats get = do
  (count, pids, counts, sizes) <- invoke get
  usingMalloced pids $
    usingMalloced counts $
    usingMalloced sizes $
    forM [0 .. fromIntegral count - 1] $ \i -> do
      pid <- peekElemOff pids i
      count <- peekElemOff counts i
      size <- peekElemOff sizes i
      return (Pid pid, PredicateStats (fromIntegral count) (fromIntegral size))
